src/OptimalFilter.jl

```diff
@@ -0,0 +1,504 @@
+    # Diagonal(Lambda) * extended_array[:], but avoids flattening and reshaping extended_array.
+    normalized_inverse_2d_fft!(extended_array, reshape(offline_matrices.Lambda, 2*params.nx, 2*params.ny).*extended_array, params)
+
+    transformed_array .= real.(@view(extended_array[1:params.nx+1, 1:params.ny+1]))
+
+end
+
+# Get a vector of values of field at positions of [stations.ist, stations.jst]
+function get_values_at_stations(field::AbstractMatrix{T}, stations::StationVectors) where T
+
+    values = zeros(T, length(stations.ist))
+
+    for (num,(i,j)) in enumerate(zip(stations.ist, stations.jst))
+        values[num] = field[i,j]
+    end
+
+    return values
+
+end
+
+function init_offline_matrices(params::TestParameters, stations::StationVectors)
+
+    n1 = (params.nx + 1) * (params.ny + 1) # number of elements in original grid
+    n1_bar = 4 * params.nx * params.ny     # number of elements in extended grid
+
+    F = Float64
+    C = ComplexF64
+
+    matrices = OfflineMatrices(Vector{F}(undef, n1_bar),                   #rho_bar
+                               Matrix{F}(undef, n1, params.nobs),          #R12
+                               Matrix{F}(undef, params.nobs, n1_bar),      #R21_bar
+                               Matrix{F}(undef, params.nobs, params.nobs), #R22
+                               Matrix{F}(undef, params.nobs, params.nobs), #R22_inv
+                               Matrix{F}(undef, n1, params.nobs),          #R12_invR22
+                               Vector{F}(undef, n1_bar),                   #Lambda (diagonal elements)
+                               Matrix{C}(undef, params.nobs, n1_bar),      #K
+                               Matrix{F}(undef, params.nobs, params.nobs), #L
+                               Matrix{F}(undef, params.nobs, params.nobs), #mu20
+
+                               Matrix{F}(undef, params.nx+1, params.ny+1), #buf1
+                               Matrix{F}(undef, params.nx+1, params.ny+1)  #buf2
+                               )
+
+    first_column_covariance_extended_grid!(matrices.rho_bar, params)
+    covariance_stations_grid!(matrices.R12, params, stations)
+    covariance_stations_extended_grid!(matrices.R21_bar, params, stations)
+    covariance_stations!(matrices.R22, params, stations)
+    matrices.R22_inv .= inv(matrices.R22)
+    matrices.R12_invR22 .= matrices.R12 * matrices.R22_inv
+
+    fourier_coeffs = Vector{ComplexF64}(undef, n1_bar)
+    normalized_inverse_2d_fft!(fourier_coeffs, matrices.rho_bar, params)
+    matrices.Lambda .= sqrt(4 * params.nx * params.ny) * real.(fourier_coeffs)
+
+    WHbar_R12 = Matrix{C}(undef, n1_bar, params.nobs)
+    for i in 1:params.nobs
+        normalized_2d_fft!(@view(WHbar_R12[:,i]), @view(matrices.R21_bar[i,:]), params)
+    end
+    KH = Diagonal(matrices.Lambda)^(-1/2)*WHbar_R12
+    matrices.K .= KH'
+
+    A = real.(matrices.R22 - matrices.K*KH)
```

Comment:
Some more dots, for good measure
```suggestion
    A = real.(matrices.R22 .- matrices.K.*KH)
```

Reply:
Oops, K*KH is a real matrix product

Reply:
Ugh :disappointed: :facepalm: